test/groovy/PiperTestBase.groovy

```diff
@@ -0,0 +1,51 @@
+import com.lesfurets.jenkins.unit.BasePipelineTest
+
+import static ProjectSource.projectSource
+import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
+
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+
+public class PiperTestBase extends BasePipelineTest {
+
+    @Rule
+    public TemporaryFolder pipelineFolder = new TemporaryFolder()
+
+    private File pipeline
+
+    protected messages = [], shellCalls = []
+
+    protected final void _setUp() {
```

Comment:
Why don't you use @Before, anyways is always called from a method annotated with @Before, so it is not necessary to write the call to this method in the upper methods, and you can name it setUp()?

Reply:
Having a setup method in a base class annotated with @Before and having a setup method in a derived class also annotated with @Before will - to my knowledge - end up in having the setup method in the derived class called twice without calling the setup method in the base class.

The thing with the slightly different name is only a workaround since calling the super class setup method (```super.setup()```)  results in a call of ```this.setup()```. Which is wrong IMO, but this is like it is.

In general I prefer having the setUp() in the base class and setUp() in the derived test classes. The call to ```super.setUp()``` (as the first call in the derived ```setUp()``` ) comes also with the nice side effect that a developer has more situation awareness compared with a situation where something is done more automatically.

Reply:
"Having a setup method in a base class annotated with @before and having a setup method in a derived class also annotated with @before will - to my knowledge - end up in having the setup method in the derived class called twice without calling the setup method in the base class."
I think this is wrong, first the setUp() method of the base class would be called, and after the setUp() method of the derived class would be called. Just like you do it, but with less code.

"The thing with the slightly different name is only a workaround since calling the super class setup method (super.setup()) results in a call of this.setup(). Which is wrong IMO, but this is like it is."
Yes this is like it is, this was already clear for me, but you don't need the workaround if you do it as I did suggest.

"In general I prefer having the setUp() in the base class and setUp() in the derived test classes. The call to super.setUp() (as the first call in the derived setUp() ) comes also with the nice side effect that a developer has more situation awareness compared with a situation where something is done more automatically."
If we would follow your situation awareness approach, we would not write clean code.